libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,693 @@
+        // `Signature` should be generated by calling `Signature::new_with_rk_mitigation`
+        // to avoid rogue key attacks. If using proof of possession mitigation
+        // then `Signature` can be generated by calling `Signature::new`
+        pub fn new(signatures: &[Signature]) -> Self {
+            AggregatedSignature(signatures.iter().fold(G2::identity(), |sig, s| sig + &s.0))
+        }
+
+        // Verify with rogue key attack mitigation.
+        pub fn verify(&self, message: &[u8], pk: &AggregatedPublicKey) -> bool {
+            let point = G2::from_msg_hash(message);
+            GT::ate_2_pairing(&-&G1::generator(), &self.0, &pk.0, &point).is_one()
+        }
+
+        // Verify without rogue key mitigation. Assumes caller has handled
+        // rogue key mitigation some other way like proof of possession.
+        // This practice is discouraged in favor of the other method
+        // but there are use cases where proof of possession is better suited
+        pub fn verify_no_rk(&self, message: &[u8], pks: &[PublicKey]) -> bool {
+            let apk = pks.iter().fold(G1::identity(), |a, p| a + &p.0);
+            let point = G2::from_msg_hash(message);
+            GT::ate_2_pairing(&-&G1::generator(), &self.0, &apk, &point).is_one()
+        }
+
+        /// This should be used to verify quickly multiple BLS aggregated signatures by batching
+    /// versus verifying them one by one as it reduces the number of computed pairings
+        pub fn batch_verify(inputs: &[(&[u8] /* message */, &AggregatedSignature, &AggregatedPublicKey)]) -> bool {
+            // To combat the rogue key attack and avoid checking for distinct messages
+            // use batch verification as described in the end of section 3.1 from https://eprint.iacr.org/2018/483
+            let mut pairs = Vec::new();
+            let mut sig = G2::identity();
+            for (msg, asg, apk) in inputs {
+                let random_exponent = FieldElement::random();
+                let g2 = G2::from_msg_hash(msg);
+                sig += &asg.0 * &random_exponent;
+                pairs.push((&apk.0 * &random_exponent, g2));
+            }
+
+            pairs.push((-&G1::generator(), sig));
+
+            let ate_pairs = pairs.iter().map(|(g1, g2)| (g1, g2)).collect();
+            GT::ate_multi_pairing(ate_pairs).is_one()
+        }
+    }
+
+    /// Creates a new BLS key pair
+    pub fn generate() -> (PublicKey, PrivateKey) {
+        let sk = PrivateKey::random();
+        let pk = PublicKey::new(&sk);
+        (pk, sk)
+    }
+
+    bls_tests_impl!();
+}
+
+/// This version is the small BLS signature scheme
+/// with the public key group in G2 and signature group in G1.
+/// 97 bytes signatures and 192 byte public keys
+///
+/// This results in smaller signatures but slower operations and bigger public key.
+/// This is good for situations where space is a consideration and verification is infrequent
+pub mod small {
+    use super::*;
```

Comment:
My comments from `mod normal` apply here as well as the code here is same as in `mod normal`